src/api_sketch.jl

```diff
@@ -0,0 +1,211 @@
+          to the associated onehot bin
+
+FEATURIZATION: the process of attaching encoded features to an atoms object (in a 
+               slight abuse of terminology, a featurization function  may also perform
+               feature encoding).
+
+File organization: probably eventually will be cleanest to have a `feat` folder with 
+separate files defining each feature type and similarly for the atoms object types and
+featurizations.
+
+In addition, there should be some place to drop functions for automatically building 
+features and featurizations for commonly desired features (i.e. things from lookup table
+and things making use of that dictionary of rdkit functions).
+
+For now, I'm sketching this all out in one file though.
+=#
+
+#= FEATURE OBJECTS
+
+I decided on an abstract type to allow the dispatch shown below in the "magical encoding"
+and "magical decoding" bits.
+
+Tn is the "natural" type of feature values, Te is the type of encoded values for the WHOLE
+STRUCTURE (not e.g. for a single atom).
+
+Representative examples – feature: feature object type, Tn, Te:
+
+block (s,p,d,f):            AtomFeat (contextual=false), String, Flux.OneHotMatrix 
+electronegativity (binned): AtomFeat (contextual=false), Float32, Flux.OneHotMatrix
+electronegativity (direct): AtomFeat (contextual=false), Float32, Vector{Float32}
+oxidation state:            AtomFeat (contextual=true) Int, Vector{Float32}
+distance between atoms:     PairFeat, Float32, Matrix{Float32}
+bond type:                  PairFeat, String, Array{Float32,3}
+
+(bond type is categorical and gets one-hot encoded, so the first two indices of the Matrix
+should be atom indices and the third should be indexing into the one-hot vector, which should just contain all zeros if the two atoms are not bonded (or we add a bin to the one-hot encoding to indicate that)
+
+All subtypes should define `encode_f` and `decode_f`
+`encode_f` should take in an atoms object and return Te
+`decode_f` should take in something of type Te and return something of type Tn
+=#
+abstract type Feature{Tn,Te} end
+
+# I think we should be able to do this too...there might be a more idiomatic way
+# to do what I'm showing below, perhaps via an "actual" holy trait type thing
+isatoms(::Type{AtomGraph}) = true
+isatoms(::Type{WeaveMol}) = true
+# add more for other types later on...
+isatoms(::Type{Any}) = false
+
+# magical encoding
+function (f<:Feature{Tn,Te})(atomsobj::T) where T
+    @assert isatoms(T) "Features can only be computed on atomic structures!"
+    f.encode_f(atomsobj)
+end
+
+# magical decoding
+(f<:Feature{Tn,Te})(encoded::Te) = f.decode_f(encoded)
+
+#=
+Feature of a single atom.
+
+(logspaced field not necessary because it can be embedded in encode_f and decode_f through 
+keywords to constructor (see below))
+=#
+struct AtomFeat{Tn,Te}<:Feature{Tn,Te}
+    name # symbol (for easy DataFrame indexing) or String (for easy other things)?
```

Comment:
anyone have thoughts/feelings about this?

Reply:
Do you think we would want pair features to be built by composing atom features with other fields?

Reply:
not usually...I think most of the relevant pair features are things like bond type where you can't really compute it easily from individual atom features